### ISSUE TITLE
fcn score.py

### ISSUE THREAD

**Issue by ZoneFv** [ZoneFv/FCN#10]
from **future** import division
import caffe
import numpy as np
import os
import sys
from datetime import datetime
from PIL import Image
import string

def fast_hist(a, b, n):
    k = (a >= 0) & (a < n)
    return np.bincount(n \* a[k].astype(int) + b[k], minlength=n**2).reshape(n, n)

def compute_hist(net, save_dir, dataset, layer='score', gt='label'):
    n_cl = 2 #net.blobs[layer].channels
    if save_dir:
        os.mkdir(save_dir)
    hist = np.zeros((n_cl, n_cl))
    loss = 0
    for idx in dataset:
        net.forward()

```
    arr = (1.0 / (1 + np.exp(-net.blobs[layer].data[0][0,:,:])) > 0.2)
    #print arr.shape, np.max(arr), np.min(arr)

    hist += fast_hist(net.blobs[gt].data[0, 0].flatten(),
                            arr.flatten(),
                            n_cl)

    if save_dir:
        im = Image.fromarray(arr.astype(np.uint8) * 255)#, mode='P')
        dirs = idx.split('/')
        if len(dirs) == 2 and not os.path.exists(save_dir+'/'+dirs[0]):
            os.mkdir(save_dir+'/'+dirs[0])
        im.save(os.path.join(save_dir, idx + '.png'))
    # compute the loss as well
    loss += net.blobs['loss_text'].data.flat[0]
return hist, loss / len(dataset)
```

def seg_tests(solver, save_format, dataset, layer='score', gt='label'):
    print '>>>', datetime.now(), 'Begin seg tests'
    solver.test_nets[0].share_with(solver.net)
    do_seg_tests(solver.test_nets[0], solver.iter, save_format, dataset, layer, gt)

def do_seg_tests(net, iter, save_format, dataset, layer='score', gt='label'):
    n_cl = net.blobs[layer].channels

```
#if save_format:
#    save_format = save_format.format(iter)
iter = save_format.split("_")[1]
hist, loss = compute_hist(net, save_format, dataset, layer, gt)

file = open('test_result.txt','a')
file.write("iter: "+str(iter))

# mean loss
print '>>>', datetime.now(), 'Iteration', iter, 'loss', loss
file.write(" "+str(loss))
# overall accuracy
acc = np.diag(hist).sum() / hist.sum()
print '>>>', datetime.now(), 'Iteration', iter, 'overall accuracy', acc
file.write(" "+str(acc))
# per-class accuracy
acc = np.diag(hist) / hist.sum(1)
print '>>>', datetime.now(), 'Iteration', iter, 'mean accuracy', np.nanmean(acc)
file.write(" "+str(np.nanmean(acc)))
# per-class IU
iu = np.diag(hist) / (hist.sum(1) + hist.sum(0) - np.diag(hist))
print '>>>', datetime.now(), 'Iteration', iter, 'mean IU', np.nanmean(iu)
file.write(" "+str(np.nanmean(iu)))
freq = hist.sum(1) / hist.sum()
print '>>>', datetime.now(), 'Iteration', iter, 'fwavacc', \
        (freq[freq > 0] * iu[freq > 0]).sum()
file.write(" "+str((freq[freq > 0] * iu[freq > 0]).sum())+"\n")
file.close()
return hist
```
